README.md

```diff
@@ -1,118 +1,228 @@
 
-
-This library provides a typescript implementation of the dag-jose codec for ipld.
+This library provides a TypeScript implementation of the DAG-JOSE codec for IPLD.
 
 It supports the new [multiformats](https://github.com/multiformats/js-multiformats) library in order to be compatible with both the current and future js-ipfs implementations.
 
-To create dag-jose compatible jose objects we recommend using the [dag-jose-utils](https://github.com/ceramicnetwork/js-dag-jose-utils) library.
+To create and work with DAG-JOSE compatible JOSE objects we recommend using the [dag-jose-utils](https://github.com/ceramicnetwork/js-dag-jose-utils) library.
 
-## Usage
-First setup the codec:
-```js
-import dagJose from 'dag-jose'
-import multiformats from 'multiformats/basics.js'
-import legacy from 'multiformats/legacy.js'
-import Ipfs from 'ipfs'
+* [JWS Signing Usage](#jws-signing-usage)
+* [JWE Encryption Usage](#jwe-encryption-usage)
+  * [Symmetric encryption](#symmetric-encryption)
+  * [Asymmetric encryption](#asymmetric-encryption)
+* [Maintainer](#maintainer)
+* [License](#license)
+
+## JWS Signing Usage
+
+_The following example is available in complete form in [example-signing-ipld.mjs](./example-signing-ipld.mjs)._
+
+For independent usage as an IPLD codec:
 
-multiformats.multicodec.add(dagJose)
-const format = legacy(multiformats, dagJose.name)
-const ipfs = await Ipfs.create({ ipld: { formats: [format] } })
+```js
+import * as Block from 'multiformats/block'
+import { sha256 } from 'multiformats/hashes/sha2'
+import * as dagCbor from '@ipld/dag-cbor' // for decoding the signed payload
+import * as dagJose from 'dag-jose'
 ```
 
-### JWS
-Import additional libraries for JWS creation
+Import additional libraries for JWS handling:
+
 ```js
+// JWT & utilities
 import {
-  encodePayload,
-} from 'dag-jose-utils'
-import {
-  EllipticSigner,
+  ES256KSigner,
   createJWS,
+  verifyJWS
 } from 'did-jwt'
-import * as u8a from 'uint8arrays'
+import {
+  encodePayload,
+  toJWSPayload,
+  toJWSStrings
+} from 'dag-jose-utils'
+```
+
+Given a keypair:
+
+```js
+const pubkey = '03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479'
+const privkey = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
 ```
 
+Create a signed envelope block:
+
 ```js
-// prepare signer
-const privkey = // hex private key
-const signer = new EllipticSigner(privkey)
+const signer = ES256KSigner(privkey)
+// arbitrary data to DAG-CBOR encode, we get a:
+// { cid:CID, linkedBlock: Uint8Array }
+const payloadBlock = await encodePayload(payload)
+// sign the CID as a JWS using our signer
+const jws = await createJWS(toJWSPayload(payloadBlock), signer)
+// convert our JWS to a DagJWS for IPLD encoidng
+const dagJWS = dagJose.prepare(jws)
```

Comment:
yes .. but .. there's a problem of symmetry, as an example I think it'd be better to demonstrate having properly isomorphic objects, the `dagJWS` you have here is going to be identical to the `dagJWS` you'd have if it went through the encode/decode round trip. It also opens up the possibility of deprecating and removing the sloppy handling in `encode()` like we've done with dag-pb -- which requires that you give it exactly the form of object that will be present in a round-trip, the user is responsible for handing the codec a properly formed data model object, but we can give them utilities to help them get there.

if you look at the current example in the readme you see this problem potentially popping up (I ran into it when messing around with it myself). Unless the object goes through the encode/decode then you can't use it properly. There are legimitate cases where you may never encode an object, but need to have its valid data model form.

basically - best practice is to have a clean mapping of the data model form such that it's stable through an encode/decode cycle.

but this is only an example, and maybe this needs to be explained?

Reply:
So any JOSE library will (for JWS) create an object which has three proerties: `payload`, `header`, and `signature` (or some variation of this depending on the JOSE encoding). With DagJWS we add an additional property called `link` which contains a CID instance which is derived from the payload (which is a base65url CID). This link property will only be populated when retrieving the content from the dag so that we can traverse it.

The main reason I want the codec to support multiple JOSE encodings is so that you can just put the output of any JOSE library directly into the DAG without having to think more about it.